File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeIncrementalIndexer.kt
package io.github.rajibsarwar.edgeframeworks

import kotlinx.coroutines.ensureActive
import kotlin.coroutines.coroutineContext

class EdgeIncrementalIndexer(
    private val retriever: EdgeRetriever,
    private val catalog: EdgeKnowledgeCatalog
) {
    suspend fun sync(
        sourceId: String,
        sourceIdentifier: String,
        contentFingerprint: String,
        documents: List<EdgeDocument>,
        collection: EdgeKnowledgeCollection,
        metadata: Map<String, String> = emptyMap(),
        chunker: EdgeTextChunker = EdgeTextChunker()
    ): EdgeSourceSyncResult {
        coroutineContext.ensureActive()

        val existing = catalog.source(sourceId)

        if (
            existing != null &&
            existing.collectionId == collection.id &&
            existing.contentFingerprint ==
                contentFingerprint
        ) {
            return EdgeSourceSyncResult.Unchanged(
                existing
            )
        }

        var removedChunkCount = 0

        if (existing != null) {
            for (documentId in existing.documentIds) {
                coroutineContext.ensureActive()
                removedChunkCount += retriever.remove(
                    documentId = documentId,
                    collectionId = existing.collectionId
                )
            }
        }

        for (document in documents) {
            coroutineContext.ensureActive()
            retriever.index(
                document = document,
                collection = collection,
                chunker = chunker
            )
        }

        val source = EdgeKnowledgeSource(
            id = sourceId,
            collectionId = collection.id,
            sourceIdentifier = sourceIdentifier,
            contentFingerprint = contentFingerprint,
            documentIds = documents.map { it.id },
            metadata = metadata
        )

        catalog.upsert(collection)
        catalog.upsert(source)

        return EdgeSourceSyncResult.Indexed(
            source = source,
            removedChunkCount =
                removedChunkCount,
            indexedDocumentCount =
                documents.size
        )
    }

    suspend fun removeSource(
        id: String
    ): Int {
        val source = catalog.source(id)
            ?: return 0

        var removed = 0

        for (documentId in source.documentIds) {
            coroutineContext.ensureActive()
            removed += retriever.remove(
                documentId = documentId,
                collectionId = source.collectionId
            )
        }

        catalog.removeSource(id)
        return removed
    }

    suspend fun removeCollection(
        collection: EdgeKnowledgeCollection
    ): Int {
        val removed =
            retriever.clear(collection)

        catalog.removeCollection(
            collection.id
        )

        return removed
    }
}

File: ios/Sources/EdgeFrameworks/Runtime/EdgeIncrementalIndexer.swift
import Foundation

public struct EdgeIncrementalIndexer: Sendable {
    private let retriever: EdgeRetriever
    private let catalog: any EdgeKnowledgeCatalog

    public init(
        retriever: EdgeRetriever,
        catalog: any EdgeKnowledgeCatalog
    ) {
        self.retriever = retriever
        self.catalog = catalog
    }

    public func sync(
        sourceID: String,
        sourceIdentifier: String,
        contentFingerprint: String,
        documents: [EdgeDocument],
        collection: EdgeKnowledgeCollection,
        metadata: [String: String] = [:],
        chunker: EdgeTextChunker = EdgeTextChunker()
    ) async throws -> EdgeSourceSyncResult {
        try Task.checkCancellation()

        if let existing = await catalog.source(id: sourceID),
           existing.collectionID == collection.id,
           existing.contentFingerprint == contentFingerprint {
            return .unchanged(existing)
        }

        let previous = await catalog.source(id: sourceID)
        var removedChunkCount = 0

        if let previous {
            for documentID in previous.documentIDs {
                try Task.checkCancellation()
                removedChunkCount += try await retriever.remove(
                    documentID: documentID,
                    collectionID: previous.collectionID
                )
            }
        }

        for document in documents {
            try Task.checkCancellation()
            try await retriever.index(
                document,
                in: collection,
                chunker: chunker
            )
        }

        let source = EdgeKnowledgeSource(
            id: sourceID,
            collectionID: collection.id,
            sourceIdentifier: sourceIdentifier,
            contentFingerprint: contentFingerprint,
            documentIDs: documents.map(\.id),
            metadata: metadata
        )

        try await catalog.upsert(
            collection: collection
        )
        try await catalog.upsert(
            source: source
        )

        return .indexed(
            source: source,
            removedChunkCount: removedChunkCount,
            indexedDocumentCount: documents.count
        )
    }

    public func removeSource(
        id: String
    ) async throws -> Int {
        guard let source = await catalog.source(id: id) else {
            return 0
        }

        var removed = 0

        for documentID in source.documentIDs {
            try Task.checkCancellation()
            removed += try await retriever.remove(
                documentID: documentID,
                collectionID: source.collectionID
            )
        }

        try await catalog.removeSource(id: id)
        return removed
    }

    public func removeCollection(
        _ collection: EdgeKnowledgeCollection
    ) async throws -> Int {
        let removed = try await retriever.clear(collection)
        try await catalog.removeCollection(id: collection.id)
        return removed
    }
}
